keps/4377-metrics-tls/README.md

```diff
@@ -0,0 +1,371 @@
+- ServiceMonitor will verify the TLS certificates from a metrics secret created by external certificate solution.
+
+### Unsupported Deployment Options
+
+Not supported Option:
+
+- Admin deploys Kueue and Cert Manager
+- Sets InternalCertManagement to true
+- Cert manager is used to provision certificates for metrics
+- Internal cert rotation is used for webhook.
+- ServiceMonitor will verify the TLS certificates from a metrics secret.
+
+To block this option we will validate that one cannot fall into this scenario.
+
+### Goals
+
+- Provide a way for metrics to verify tls certificates.
+
+### Non-Goals
+
+- Disable secure metrics serving.
+
+It is important to secure metrics if they are enabled.
+We will not provide a way to disable this.
+
+- Allow for external certificates for metrics but internal certificates for webhooks.
+
+- Adopt Internal Cert Manager solution for metrics and provide self signed certificates.
+
+- Supporting APIService certificates
+
+APIService currently uses Insecure flags but there has been no effort made yet to enhance that for Cert manager. It would be good to have support for this but it is out of scope of this KEP.
+
+## Proposal
+
+We will require that if one is using external certificates and prometheus is enabled,
+one must use certificates for metrics. 
+We will effectively be disabling the use of controller runtime generated certificates if an external certificate solution is used.
+
+We will pass this options to the controller runtime `MetricsServer` option where the validation will be done by
+the controller runtime.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster adminstration, I want to use an external certificate service for all certificates. 
+I want to use external certificates for webhooks and metrics.
+I am concerned with turning off tls verification for prometheus and I want to validate the certificates.
+
+### Notes
+
+Kueue at the moment skips all tls checking for prometheus endpoints.
+
+```yaml
+# Prometheus Monitor Service (Metrics)
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: controller-manager-metrics-monitor
+  namespace: system
+spec:
+  endpoints:
+    - path: /metrics
+      port: https
+      scheme: https
+      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+      tlsConfig:
+        insecureSkipVerify: true
+  selector:
+    matchLabels:
+      app.kubernetes.io/component: controller
+      app.kubernetes.io/name: kueue
+```
+
+### Risks and Mitigations
+
+This is more secure option for deploying Kueue but it would require an external dependency on Cert Manager or some other
+certificate management solution.
+
+## Design Details
+
```

Comment:
Actually, would you mind adding a clarifying section as below:

```golang
	// Enable controls the use of internal cert management for the webhook
	// and metrics endpoints. When enabled Kueue is using libraries to generate and
	// self-sign the certificates. When disabled, you need to provide the certificates for 
	// the endpoints:
        // - for webhooks the recommended approach is to use the third-party cert-manager. 
        // - for metrics the external certs are expected to be provided on a mounted volume 
        //    under the `/etc/kueue/metrics-certs` path.
        // 
        // Defaults to true.
	Enable *bool `json:"enable,omitempty"`
```

WDYT? cc @tenzen-y

Reply:
Are you saying we should update the comment here?

Reply:
Well this is just KEP, but I think it makes sense to propose the comment update in KEP as the semantics of the field is changing with this work. Previously the field only mattered for webhooks In the core k8s we also add and update API comments in keps.

Reply:
I am still not sure what you wanted me to add. I tried to take a guess at what you mean in the latest commit.

Your comment is not really correct as Cert Manager flow is the same for both webhooks and metrics. 

a) Cert Manager creates a certificate that references a secret and service.
b) Kueue controller manager mounts the secret that contains tls/keys.

These steps are the same for both metrics and webhooks. We only provide the secrets/ca for the service monitor but cert manager flow is identical for both.

Reply:
hmm, ok but in case of webhooks we inject using cainjector and for metrics we mount them as volume, which differs the flow?

Reply:
btw. if you think the comment is incorrect please adjust to make it correct. I'm still learning the stuff so I'm happy to learn. my point is that this PR changes the semantics of the field, so it is a good opportunity to update the comment. This was just an initial proposal for discussion.

Reply:
regarding the intended change I imagined just a new block with the API field and comment. Similarly as we propose new Api in other keps like tas

Reply:
I see. I added a section..

Reply:
thx

Reply:
updated.